File: cr3.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func letterValue(r rune) int {
	return int(r - 'a' + 1)
}

func extractNumber(s string) float64 {
	buf := ""
	for _, r := range s {
		if ('0' <= r && r <= '9') || r == '/' || r == '-' || r == '.' {
			buf += string(r)
		}
	}
	if strings.Contains(buf, "/") {
		parts := strings.Split(buf, "/")
		p,_ := strconv.Atoi(parts[0])
		q,_ := strconv.Atoi(parts[1])
		return float64(p)/float64(q)
	}
	n,_ := strconv.ParseFloat(buf,64)
	return n
}

func cot(x float64) float64 {
	return 1.0 / math.Tan(x)
}

func meaning(sum int) string {
	switch sum {
	case 5: return "BOX"
	case 6: return "ZAHL"
	case 7: return "WINKEL"
	case 8: return "RAUM"
	default: return "INVALID"
	}
}

func main() {
	cmd := "d;h(3;1/2,cot(1/2);0,0,0;1,0,0;p1,p2)"

	sum := letterValue(rune(cmd[0])) + letterValue(rune(cmd[2]))
	fmt.Println("// Befehl:", meaning(sum))

	if sum == 6 {
		num := extractNumber(cmd)
		fmt.Printf("value := %f\n", num)
	} else if sum == 7 {
		num := extractNumber(cmd)
		fmt.Printf("tan_theta := %f\n", math.Tan(num))
		fmt.Printf("cot_theta := %f\n", cot(num))
	} else if sum == 8 {
		dim_count := 3
		angles := []float64{0.5, cot(0.5), math.Pi/4}
		points := [][]float64{{0,0,0},{1,0,0},{0,1,0}}
		ids := []string{"p1","p2","p3"}

		fmt.Printf("// RAUM: dim_count=%d\n", dim_count)
		for i:=0;i<dim_count;i++ {
			fmt.Printf("Angle %d: tan=%f cot=%f\n", i+1, angles[i], cot(angles[i]))
			fmt.Printf("Point %d: (%f,%f,%f) id=%s\n", i+1, points[i][0], points[i][1], points[i][2], ids[i])
		}
	} else if sum == 5 {
		fmt.Println("// BOX: leerer Befehl")
	}
}

File: cr3.zig
const std = @import("std");
const math = std.math;

fn letterValue(c: u8) i32 {
    return @intCast(i32, c - 'a' + 1);
}

fn extractNumber(s: []const u8) f64 {
    var buf: [64]u8 = undefined;
    var j: usize = 0;
    for (s) |c| {
        if ((c >= '0' and c <= '9') or c == '/' or c == '-' or c == '.') {
            buf[j] = c;
            j += 1;
        }
    }
    const str = buf[0..j];
    for (str) |c| {}
    if (std.mem.indexOf(u8, str, "/") != null) {
        var parts = std.mem.split(u8, str, "/");
        const p = std.fmt.parseInt(i32, parts[0], 10) catch 0;
        const q = std.fmt.parseInt(i32, parts[1], 10) catch 1;
        return @intToFloat(f64, p)/@intToFloat(f64,q);
    }
    return std.fmt.parseFloat(f64,str,10) catch 0.0;
}

fn cot(x:f64) f64 {
    return 1.0 / math.tan(x);
}

fn meaning(sum: i32) []const u8 {
    return switch(sum){
        5 => "BOX",
        6 => "ZAHL",
        7 => "WINKEL",
        8 => "RAUM",
        else => "INVALID"
    };
}

pub fn main() void {
    const cmd = "d;h(3;1/2,cot(1/2);0,0,0;1,0,0;p1,p2)";
    const sum = letterValue(cmd[0]) + letterValue(cmd[2]);
    const num = extractNumber(cmd);

    std.debug.print("// Befehl: {s}\n", .{meaning(sum)});
    if (sum == 6) {
        std.debug.print("value: f64 = {f}\n", .{num});
    } else if (sum == 7) {
        std.debug.print("tan_theta: f64 = {f}\n", .{math.tan(num)});
        std.debug.print("cot_theta: f64 = {f}\n", .{cot(num)});
    } else if (sum == 8) {
        const dim_count: i32 = 3;
        const angles: [3]f64 = [3]f64{0.5, cot(0.5), math.pi/4};
        const points: [[3]f64] = [[3]f64{0,0,0}, [3]f64{1,0,0}, [3]f64{0,1,0}];
        const ids: [3][]const u8 = [_][]const u8{"p1","p2","p3"};

        std.debug.print("// RAUM: dim_count={d}\n", .{dim_count});
        for (angles) |angle, i| {
            std.debug.print("Angle {d}: tan={f} cot={f}\n", .{i+1, angle, cot(angle)});
            std.debug.print("Point {d}: ({f},{f},{f}) id={s}\n", .{i+1, points[i][0], points[i][1], points[i][2], ids[i]});
        }
    } else if (sum == 5) {
        std.debug.print("// BOX\n", .{});
    }
}/
